### PR TITLE
New module: Shaping rules

### PR DESCRIPTION
**Pull request by mkomorski**
This module is scoped to define which bidders should be in a given auction based on a defined set of parameters. That would save money for the host company and for the bidder by eliminating network traffic for scenarios where they aren't likely to bid. It addresses these use cases:

- Don't call specific bidders for users in particular geographies
- Don't call specific bidders when there's no EID or FPD
- General A/B testing for experimenting with individual new bidders
- General A/B testing for experimenting with sets of bidders

This feature is also of interest to the Prebid Sustainability Taskforce, since making fewer requests saves money and reduces carbon emissions.

## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] New bidder adapter  
- [ ] Updated bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes

- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Other information

#12663

### CODE REVIEW

File: src/adapterManager.ts [hunk 1]
```diff
@@ -503,49 +503,69 @@ const adapterManager = {
...
 
+    let {[PARTITIONS.CLIENT]: clientBidders, [PARTITIONS.SERVER]: serverBidders} = partitionBidders(adUnits, _s2sConfigs);
+
+    const ortb2 = ortb2Fragments.global || {};
+    const bidderOrtb2 = ortb2Fragments.bidder || {};
+
+    const getTid = tidFactory();
+
+    const mergeBidderFpd = (() => {
+      const fpdCache: Record<BidderCode, any> = {};
```
Comment (dgirardi) [inline src/adapterManager.ts]: this type is not correct, so why specify it? any is better.

File: src/adapterManager.ts [hunk 2]
```diff
@@ -503,49 +503,69 @@ const adapterManager = {
...
+      return function(auctionId: string, bidderCode: BidderCode, s2sActivityParams?) {
+        const cacheKey = bidderCode + (s2sActivityParams != null ? s2sActivityParams[ACTIVITY_PARAM_S2S_NAME] : '');
```
Comment (dgirardi) [inline src/adapterManager.ts]: Simple string concat is risky here, I could have bidders `a`, `ab` and an s2s named `b`.

File: src/adapterManager.ts [hunk 3]
```diff
@@ -503,49 +503,69 @@ const adapterManager = {
       .filter(uniques)
       .forEach(incrementAuctionsCounter);
 
+    let {[PARTITIONS.CLIENT]: clientBidders, [PARTITIONS.SERVER]: serverBidders} = partitionBidders(adUnits, _s2sConfigs);
```
Comment (dgirardi) [inline src/adapterManager.ts]: I think this wouldn't give the same result now that it happens before the filtering.

File: modules/rules/index.ts [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+        continue;
+      }
+      registerResult();
+    }
+  }
+}
+
+export function evaluateSchema(func, args, context) {
+  switch (func) {
+    case 'percent':
```
Comment (dgirardi) [inline modules/rules/index.ts]: An old fashioned map would be more readable (and less bug -prone) than a `switch` IMO

File: test/spec/modules/rules_spec.js [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+        }]
+      };
+
+      sandbox.stub(Math, 'random').returns(0.5);
+      rulesModule.evaluateConfig(rulesJson);
+
+      // Verify that excluded bidder is not allowed for processed-auction
+      expect(isActivityAllowed(ACTIVITY_ADD_BID_RESPONSE, activityParams(MODULE_TYPE_BIDDER, 'bidder3', {}))).to.eql(false);
+      // Verify that non-excluded bidder is allowed for processed-auction
+      expect(isActivityAllowed(ACTIVITY_ADD_BID_RESPONSE, activityParams(MODULE_TYPE_BIDDER, 'bidder4', {}))).to.eql(true);
```
Comment (dgirardi) [inline test/spec/modules/rules_spec.js]: I am not seeing how you're passing in all the extra context required by some rules (`adUnit`, `regs`, `ortb2`, etc), so I was looking for tests about them, and noticed that all the tests requiring it are not calling `isActivityAllowed` - so I'm guessing that part is still missing?

File: test/spec/modules/rules_spec.js [hunk 6]
```diff
@@ -0,0 +1,568 @@
...
+      const context3 = {
+        ortb2: {}
+      };
+      const func3 = rulesModule.evaluateSchema('fpdAvailable', [], context3);
+      expect(func3()).to.be.false;
+    });
+
+    it('should evaluate gppSidIn condition', function() {
+      const context1 = {
+        regs: {
```
Comment (dgirardi) [inline test/spec/modules/rules_spec.js]: why is regs outside of `ortb2` for this rule?

File: test/spec/modules/rules_spec.js (hunk 5 above)
Comment (mkomorski) [reply test/spec/modules/rules_spec.js]: [here's](https://github.com/prebid/Prebid.js/pull/14079/files#diff-892b70a36d919d6bc589b1d541fecd97fd66b3cf5c3ca9df9c6e66ad46d1028fR551) where it's passed on bid request phase but you're right, it's missed to pass it to the response phase

Comment (mkomorski) [reply test/spec/modules/rules_spec.js]: These `describe('evaluateSchema')` are unit tests limited only to check if [schema functions](https://docs.google.com/document/d/1LcUkGfWL4aKFys0MRSNjKgo8NML2gS3VUtf1GQFKF_4/edit?tab=t.0#heading=h.g810jmahah48) return correct value for given params `function`, `args`, `context` (where `context` holds adUnit, bid, etc) so they don't register any activities. 

I'm testing if activities are properly registered in `describe('evaluateConfig')` suite, and since it works correct for example schema:
`schema: [{ function: 'adUnitCode', args: ['adUnit-0000'] }]`
`const bidder1Params = activityParams(MODULE_TYPE_BIDDER, 'bidder1', {
          adUnit: { code: 'adUnit-0000' }
});`
and all the schema functions unit tests also work correct, I didn't implement "end-to-end" tests for all possible schemas functions.

Comment (mkomorski) [reply test/spec/modules/rules_spec.js]: Here's where parameters are passed:
[adapterManager.ts](https://github.com/prebid/Prebid.js/pull/14079/files#diff-892b70a36d919d6bc589b1d541fecd97fd66b3cf5c3ca9df9c6e66ad46d1028fR555)
[auction.ts](https://github.com/prebid/Prebid.js/pull/14079/files#diff-ad32dc5253fb625446a09d8f5fc488a6167880eb280f340b651e2636652f6534R462)

**Comment by patmmccann** [comment]
@mkomorski can we start the docs pr?

**Comment by monis0395** [comment]
@mkomorski 

1. Should we allow caching of rules, similar to how `topicsFpdModule` handles it? This would help prevent auction delays when the endpoint is slow or unavailable, and reduce unnecessary network requests on subsequent page loads.
2. Should we add way to pass a js functions in `schema.[].function` which will allow them to write their own rules ?
3. The example rules.JSON shows excludeModules but it's not implemented, do we plan to add in future releases ?
4. Currently, the `percent` function evaluates `Math.random()` on every check, which means the same bidder could be included or excluded inconsistently across different checks within the same auction. Should this be deterministic per auction or per user? For user-level consistency, we could generate a global random number once per session and use that as a seed.

**Comment by mkomorski** [comment]
> @mkomorski
> 
> 1. Should we allow caching of rules, similar to how `topicsFpdModule` handles it? This would help prevent auction delays when the endpoint is slow or unavailable, and reduce unnecessary network requests on subsequent page loads.
> 2. Should we add way to pass a js functions in `schema.[].function` which will allow them to write their own rules ?
> 3. The example rules.JSON shows excludeModules but it's not implemented, do we plan to add in future releases ?
> 4. Currently, the `percent` function evaluates `Math.random()` on every check, which means the same bidder could be included or excluded inconsistently across different checks within the same auction. Should this be deterministic per auction or per user? For user-level consistency, we could generate a global random number once per session and use that as a seed.

Thank you for review @monis0395 

1. we agreed on we don't make rules sticky
2. I don't think so, that was not in the scope
3. removed from the example
4. nice catch! Implemented global random

**Comment by monis0395** [comment]
> Thank you for review @monis0395
> 
>1. we agreed on we don't make rules sticky
>2. I don't think so, that was not in the scope
>3. removed from the example
>4. nice catch! Implemented global random

1. Make sense, both ways are fine.
2. Okay, we can think about it in future
3. Nice, there another example like that `buyeruidAvailable`
4. Perfect!

**Comment by robertrmartinez** [comment]
Hi all @mkomorski @monis0395 

Regarding the Math.random `random` stuff I actually would think we would want this to be evaluated every auction!

That is how PBS does it if I am not mistaken.

I get both reasonings, so I feel like maybe we should do both and make it configurable?

I have been asked by our data science team before to make decisions as granular as possible, per auction so that is my reasoning for asking.

**Comment by monis0395** [comment]
@robertrmartinez 

I am fine with using per auction level too, as long as it is not generated at every request level.

- user level
  - auction 1
    - ad unit 1 -  bidder 1 - request 1 
    - ad unit 2 - bidder 1 - request 2
  - auction 2
    - ad unit 1 -  bidder 1 - request 3 
    - ad unit 2 - bidder 1 - request 4
    
Here, the `random` value can be at the user level or the auction level. For auction level, auction 1 and auction 2 can have different `random` values. But generating it at the request level doesn't make sense.

**Comment by robertrmartinez** [comment]
Yes sorry if I worded incorrectly

I agree, its fine to just get the random number once per auction and use when selecting each rule.

File: modules/rules/index.ts [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+        logError(`${MODULE_NAME}: Unknown result function ${result.function}`);
+        continue;
+      }
+      registerResult();
+    }
+  }
+}
+
+const schemaEvaluators = {
+  percent: (args, context) => () => dep.getGlobalRandom() * 100 < args[0],
```
Comment (robertrmartinez) [inline modules/rules/index.ts]: is this not always using one `math.random` per page load?

I was thinking each auction would be able to select a new random? Sorry If I am missing something here?

**Comment by monis0395** [comment]
> Yes sorry if I worded incorrectly
> 
> I agree, its fine to just get the random number once per auction and use when selecting each rule.

Yes! Also if it is configurable at the user level or the auction level, that would be great @robertrmartinez @mkomorski

File: modules/rules/index.ts [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+  }
+
+  if (!config.enabled) {
+    logInfo(`${MODULE_NAME}: Rules engine is disabled in the configuration.`);
+    return;
+  }
+
+  const stageRules = config.ruleSets;
+
+  assignModelGroups(stageRules || []);
```
Comment (robertrmartinez) [inline modules/rules/index.ts]: so we are picking one model per page load.

I have a similar comment on this being I would think this would want to be random per auction as to get the most uniform spread of model distributions.

PBS being stateless does everything new each auction.

I guess this becomes a bit murky here because we are just hooking into auction via activityControls which is not per auction rather per bidder.

Curious on others thoughts on this.

Comment (robertrmartinez) [reply modules/rules/index.ts]: We could perhaps use the existing `requestBidsHook` to optionally run assignModelGroups each time?

Though I do not think auctionId is guaranteed to be generated at this point yet so using that to offset model selected may not work.

And if you run auctions back to back things may get weird with analytics if we just run assignModelGroups without some sort of assignment

**Comment by pm-kapil-tuptewar** [comment]
@mkomorski I’m assuming the country value will be set by the publisher through the ortb2 setConfig API so we can run country-based rules. But if a third-party rules provider creates rules that depend on country and the publisher forgets to pass it, how common is that? Would we need to rely on third-party APIs to fill in the country field in ortb2?

Also, do you think the browser should be another parameter we can use in rules — for example, including or excluding a bidder based on the user’s browser?

Lastly, regarding the bidPrice rule — I assume it removes bids below a certain price. But since the main goal of this module is to exclude bidders before sending them requests. Would love to hear your thoughts.

**Comment by mkomorski** [comment]
> @mkomorski I’m assuming the country value will be set by the publisher through the ortb2 setConfig API so we can run country-based rules. But if a third-party rules provider creates rules that depend on country and the publisher forgets to pass it, how common is that? Would we need to rely on third-party APIs to fill in the country field in ortb2?
> 
> Also, do you think the browser should be another parameter we can use in rules — for example, including or excluding a bidder based on the user’s browser?
> 
> Lastly, regarding the bidPrice rule — I assume it removes bids below a certain price. But since the main goal of this module is to exclude bidders before sending them requests. Would love to hear your thoughts.

@pm-kapil-tuptewar This is a basic version of the module based on what was implemented in the Prebid Server. I think it would be best to release it now and then actively extend it based on what users request in terms of new functionalities.

Regarding `bidPrice`, there are two stages you can choose from: `processed-auction-request` and `processed-auction` the first is excluding bidders before fetching, and the second is excluding bidders after receiving responses. `bidPrice` can be used in the second stage.

**Comment by pm-kapil-tuptewar** [comment]
thanks @mkomorski for the details.

File: modules/rules/index.ts [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+
+const delayedAuctions = timeoutQueue();
+
+interface ModuleConfig {
+  endpoint?: {
+    url: string;
+    method: string;
+  };
+  auctionDelay?: number;
+}
```
Comment (pm-komal-kumari) [inline modules/rules/index.ts:49]: Hi @mkomorski, similar to the priceFloors and timeoutRtdProvider modules, which accept both static configuration data and endpoint-based data, can we adopt a comparable structure within shapingRules module as well?

File: modules/rules/index.ts [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+    }
+  }
+}
+
+const schemaEvaluators = {
+  percent: (args, context) => () => {
+    const auctionId = context.bid?.auctionId;
+    return dep.getGlobalRandom(auctionId) * 100 < args[0]
+  },
+  adUnitCode: (args, context) => () => context.adUnit.code === args[0],
```
Comment (pm-komal-kumari) [inline modules/rules/index.ts]: Hi @mkomorski, while testing with the above example, I observed that this condition consistently evaluates to false.
The conditions like context.adUnit.code === args[0] or context.ortb2?.device?.geo?.country === args[0] are consistently evaluating to false because context.adUnit.code and context.ortb2.device.geo.country are strings, whereas args[0] is an object, resulting in a data type mismatch. Please let me know if I may have misunderstood something.

Comment (mkomorski) [reply modules/rules/index.ts]: Hi @pm-komal-kumari. Im not sure why you assume args[0] to be an object. In current shape it should be string, however I double checked with the prebid server implementation and in fact functions like adUnitCode should just return the value and comparision value should be took from conditions, so I updated the code. Please check the newest json file

File: modules/rules/index.ts (hunk 9 above)
Comment (robertrmartinez) [reply modules/rules/index.ts:49]: +1 to this we should support both

**Comment by robertrmartinez** [comment]
another thing is can we provide

> > @mkomorski I’m assuming the country value will be set by the publisher through the ortb2 setConfig API so we can run country-based rules. But if a third-party rules provider creates rules that depend on country and the publisher forgets to pass it, how common is that? Would we need to rely on third-party APIs to fill in the country field in ortb2?
> > Also, do you think the browser should be another parameter we can use in rules — for example, including or excluding a bidder based on the user’s browser?
> > Lastly, regarding the bidPrice rule — I assume it removes bids below a certain price. But since the main goal of this module is to exclude bidders before sending them requests. Would love to hear your thoughts.
> 
> @pm-kapil-tuptewar This is a basic version of the module based on what was implemented in the Prebid Server. I think it would be best to release it now and then actively extend it based on what users request in terms of new functionalities.
> 
> Regarding `bidPrice`, there are two stages you can choose from: `processed-auction-request` and `processed-auction` the first is excluding bidders before fetching, and the second is excluding bidders after receiving responses. `bidPrice` can be used in the second stage.

regarding other schemaEvaluators, could we add in a hook to add custom schemaValidators?

Similar to additionalSchemaFields in the priceFloors module?

```js
pbjs.setConfig({
  shapingRules: {
    // config stuff
    extraSchemaEvaluators: {
      browser: (args, context) => {
        // some fancy code to return browser string
      }
    }
  }
});
```

**Comment by pm-kapil-tuptewar** [comment]
> another thing is can we provide
> 
> > > @mkomorski I’m assuming the country value will be set by the publisher through the ortb2 setConfig API so we can run country-based rules. But if a third-party rules provider creates rules that depend on country and the publisher forgets to pass it, how common is that? Would we need to rely on third-party APIs to fill in the country field in ortb2?
> > > Also, do you think the browser should be another parameter we can use in rules — for example, including or excluding a bidder based on the user’s browser?
> > > Lastly, regarding the bidPrice rule — I assume it removes bids below a certain price. But since the main goal of this module is to exclude bidders before sending them requests. Would love to hear your thoughts.
> > 
> > 
> > @pm-kapil-tuptewar This is a basic version of the module based on what was implemented in the Prebid Server. I think it would be best to release it now and then actively extend it based on what users request in terms of new functionalities.
> > Regarding `bidPrice`, there are two stages you can choose from: `processed-auction-request` and `processed-auction` the first is excluding bidders before fetching, and the second is excluding bidders after receiving responses. `bidPrice` can be used in the second stage.
> 
> regarding other schemaEvaluators, could we add in a hook to add custom schemaValidators?
> 
> Similar to additionalSchemaFields in the priceFloors module?
> 
> ```js
> pbjs.setConfig({
>   shapingRules: {
>     // config stuff
>     extraSchemaEvaluators: {
>       browser: (args, context) => {
>         // some fancy code to return browser string
>       }
>     }
>   }
> });
> ```

Yes 👍

File: modules/rules/index.ts (hunk 8 above)
Comment (mkomorski) [reply modules/rules/index.ts]: I think it's possible to achieve both calculating and assigning model groups on every auction and use activity controls by running the whole current logic in startAuction hook, pass auctionId to activity controls rules params and then unregister the rules on auction end. If this is how it works on pbs, we could just do the same, without making it optional. Also curious on others opinion on that

File: modules/rules/index.ts (hunk 7 above)
Comment (dgirardi) [reply modules/rules/index.ts]: I'd use a `WeakMap` with the keys being the auctions from `auctionManager.index.getAuction`

File: modules/rules/index.ts [hunk 11]
```diff
@@ -0,0 +1,416 @@
...
+    const auctionId = context.auctiondId || context.bid?.auctionId;
+    return dep.getGlobalRandom(auctionId) * 100 < args[0]
+  },
+  adUnitCode: (args, context) => () => context.adUnit.code,
+  adUnitCodeIn: (args, context) => () => args[0].includes(context.adUnit.code),
+  deviceCountry: (args, context) => () => context.ortb2?.device?.geo?.country,
+  deviceCountryIn: (args, context) => () => args[0].includes(context.ortb2?.device?.geo?.country),
+  channel: (args, context) => () => {
+    const channel = context.ortb2?.ext?.prebid?.channel;
+    if (channel === 'pbjs') return 'web';
```
Comment (dgirardi) [inline modules/rules/index.ts]: AFAIK this is not standard. the PBS adapter will set channel = 'pbjs', some other adapters may, but in general I don't think it can be relied upon. 
Also, unless the publisher sets it (which seems unlikely), it would only work on the response side - since on the request the adapters that do set it haven't run yet.

I wonder if we could always return 'web' here? If this is taken from the PBS implementation I think it's meant to discriminate requests from pbjs from the rest, so in our case we'd always be web.

File: modules/rules/index.ts [hunk 12]
```diff
@@ -0,0 +1,416 @@
...
+    case 'excludeBidders':
+    case 'includeBidders':
+      return () => {
+        const activity = {
+          'processed-auction-request': ACTIVITY_FETCH_BIDS,
+          'processed-auction': ACTIVITY_ADD_BID_RESPONSE
+        }[stage];
+        args.forEach(({bidders, analyticsValue}) => {
+          const unregister = registerActivityControl(activity, MODULE_NAME, (params) => {
+            if ((params.auctionId || params.bid?.auctionId) !== auctionId) return { allow: true };
```
Comment (dgirardi) [inline modules/rules/index.ts]: this (and others) should just return undefined. {allow: true} means that the activity should be allowed (unless higher priority rules disallow it), undefined means leave the decision to other rules.

File: modules/rules/index.ts [hunk 13]
```diff
@@ -0,0 +1,416 @@
...
+  switch (func) {
+    case 'excludeBidders':
+    case 'includeBidders':
+      return () => {
+        const activity = {
+          'processed-auction-request': ACTIVITY_FETCH_BIDS,
+          'processed-auction': ACTIVITY_ADD_BID_RESPONSE
+        }[stage];
+        args.forEach(({bidders, analyticsValue}) => {
+          const unregister = registerActivityControl(activity, MODULE_NAME, (params) => {
```
Comment (dgirardi) [inline modules/rules/index.ts]: It'd be preferable (I think) to have just one rule that is always active (if the module is active) instead of registering / unregistering several for each auction. 

e.g. you could have a weak map from the auction object to `{excludeBidders: [args, schema, conditions, stage], includeBidders: [/* same */]}`. `evaluateFunction` would add its parameters to the map, and a rule that's set up only twice (once per activity) can look up the relevant parameters and make the decision.

The advantage compared to this is - you don't need to worry about cleaning up (I'm not sure that bidsBackCallback is the right place for it, you can theoretically still get bids after that runs); and you don't have as much overhead in the case when there are no shaping rules to run for an auction.

File: modules/rules/index.ts (hunk 11 above)
Comment (robertrmartinez) [reply modules/rules/index.ts]: yeah this should just return web if we need it at all. I think the entity providing the rules _should_ be able to discern that putting it in pbjs config means always web.

But maybe its fine to just leave it as channel: () => 'web' just in case

File: modules/rules/index.ts [hunk 14]
```diff
@@ -0,0 +1,426 @@
...
+interface RulesConfig {
+  version: string;
+  ruleSets: RuleSet[];
+  timestamp: string;
+  enabled: boolean;
+}
+
+declare module '../../src/config' {
+  interface Config {
+    shapingRules?: ShapingRulesConfig;
```
Comment (dgirardi) [inline modules/rules/index.ts:148]: I added this to make the type part of the public interface types (`pbjs.setConfig / getConfig`), also renamed `ModuleConfig` to be more meaningful in there.  It'd be nice to also use the field descriptions from https://github.com/prebid/prebid.github.io/pull/6359/files as doc comments for each property.

File: test/spec/modules/rules_spec.js [hunk 15]
```diff
@@ -0,0 +1,640 @@
...
+          user: {
+            eids: []
+          }
+        }
+      };
+      const func2 = rulesModule.evaluateSchema('eidAvailable', [], context2);
+      expect(func2()).to.be.false;
+    });
+
+    it('should evaluate userFpdAvailable condition', function() {
```
Comment (dgirardi) [inline test/spec/modules/rules_spec.js:563]: could you add a couple test cases for "when an includeBidders/excludeBidders rule is set, only the correct bidders should pass the activity check" ?

File: modules/rules/index.ts [hunk 16]
```diff
@@ -0,0 +1,491 @@
...
+    );
+
+    let randomValue = Math.random() * weightSum;
+
+    for (const group of modelGroups) {
+      group.selected = false;
+      // 100 is default weight if not specified
+      const groupWeight = group.weight ?? 100;
+      if (randomValue < groupWeight) {
+        group.selected = true;
```
Comment (dgirardi) [inline modules/rules/index.ts]: If the groups have weights 80 and 20 in order, this would select the first group 80% of the time, and no group 20% of the time.
I'm also not sure about using ".selected = true" here, what happens when you have concurrent auctions?

File: modules/rules/index.ts [hunk 17]
```diff
@@ -0,0 +1,491 @@
...
+        break;
+      }
+      randomValue -= groupWeight;
+    }
+
+    if (!modelGroups.some(g => g.selected)) {
+      modelGroups[modelGroups.length - 1].selected = true;
```
Comment (dgirardi) [inline modules/rules/index.ts]: actually, I hadn't noticed this; updated example: with weights 50, 30, 20, the first group is selected 50% of the time,  and the third group 50% of the time.

File: modules/rules/index.ts [hunk 18]
```diff
@@ -0,0 +1,506 @@
...
+
+    const weightSum = cumulativeWeight;
+    // Generate random value in range [0, weightSum)
+    // This ensures each group gets probability proportional to its weight
+    const randomValue = Math.random() * weightSum;
+
+    // Find first group where cumulative weight >= randomValue
+    let selectedIndex = groupsWithCumulativeWeights.findIndex(({ cumulativeWeight }) => randomValue < cumulativeWeight);
+
+    // Fallback: if no group was selected (shouldn't happen, but safety check)
```
Comment (dgirardi) [inline modules/rules/index.ts:240]: throwing an error would be better, or adding test cases for the edge cases and removing this even more better.

**Review by patmmccann — APPROVED** [review]
